services/src/main/java/org/keycloak/broker/saml/mappers/UserAttributeMapper.java

```diff
@@ -99,61 +101,84 @@ public String getDisplayType() {
     @Override
     public void preprocessFederatedIdentity(KeycloakSession session, RealmModel realm, IdentityProviderMapperModel mapperModel, BrokeredIdentityContext context) {
         String attribute = mapperModel.getConfig().get(USER_ATTRIBUTE);
-        String value = getAttribute(mapperModel, context);
-        if (value != null) {
-            if (attribute.equalsIgnoreCase("email")) {
-                context.setEmail(value);
-            } else if (attribute.equalsIgnoreCase("firstName")) {
-                context.setFirstName(value);
-            } else if (attribute.equalsIgnoreCase("lastName")) {
-                context.setLastName(value);
+        Map<String, List<String>> attributeToValues = getAttribute(mapperModel, context);
+        if (!attributeToValues.isEmpty()) {
+            String attributeName = getAttributeName(mapperModel);
+            if (attribute.equalsIgnoreCase(EMAIL)) {
+                setIfNotEmpty(context::setEmail, attributeToValues.get(attributeName));
+            } else if (attribute.equalsIgnoreCase(FIRST_NAME)) {
+                setIfNotEmpty(context::setFirstName, attributeToValues.get(attributeName));
+            } else if (attribute.equalsIgnoreCase(LAST_NAME)) {
+                setIfNotEmpty(context::setLastName, attributeToValues.get(attributeName));
             } else {
-                context.setUserAttribute(attribute, value);
+                context.setUserAttribute(attribute, attributeToValues.get(attributeName));
             }
         }
     }
 
-    protected String getAttribute(IdentityProviderMapperModel mapperModel, BrokeredIdentityContext context) {
+    private String getAttributeName(IdentityProviderMapperModel mapperModel) {
```

Comment:
The name obtained in this method and in getAttribute (lines 135-139) are generally different (consider an empty name or name like `  email  `).